### ISSUE TITLE
Support the new "Forwarded" header (RFC 7239)

### ISSUE THREAD

**Issue by cwe1ss** [aspnet/BasicMiddleware#28]
There is now a standard for the most common X-Forwarded-* headers. It was introduced with RFC 7239: "[Forwarded HTTP Extension](https://tools.ietf.org/html/rfc7239)". 

Example: `Forwarded: for=192.0.2.60;proto=http;by=203.0.113.43`

It would be great if this could be supported.

**Comment by Tratcher** [comment]
Yes, we've seen it. Are you aware of any proxies that actually set that new header?

**Comment by cwe1ss** [comment]
No - but I haven't looked either. I recently built a small ASP.NET gateway
proxy for Azure Service Fabric and that's why I stumbled upon this RFC. I
guess this definitely isn't high priority since there won't be anyone who's
setting just this header for a looong time. It would just be good to
support it someday since it is a standard and will eventually get more
common.

2016-01-27 17:59 GMT+01:00 Chris R <notifications@github.com>:

**Comment by lantrix** [comment]
I have a scenario whereby we have a reverse proxy (internally) passing requests to backend api’s and web apps; and one of the api's is ASP.NET 5 and using the BasicMiddleware classes to handle the headers from the upstream proxies. Of course; the team had implemented the newer RFC so BasicMiddleware can't handle `Forwarded` headers from the [RFC](https://tools.ietf.org/html/rfc7239).

**Comment by asbjornu** [comment]
We're using `Forwarded` and plan to implement it in the proxies we use, [such as NGINX](https://www.nginx.com/resources/wiki/start/topics/examples/forwarded/). The header is becoming more mainstream and it would be good of Microsoft to advocate for its use over the `X-Forwarded-*` headers.

**Comment by damianh** [comment]
Would also like to have ability to extend the with a `PathBase` value as well as support for that with the current implementation i.e. `X-Forwarded-PathBase`

Related: https://github.com/aspnet/Hosting/issues/1120

**Comment by Tratcher** [comment]
@damianh do you expect PathBase to be different per request? That Hosting issue is about getting one value at startup.

**Comment by damianh** [comment]
Yes, I would certainly like to be able to do so on a per-request basis. This will allow two scenarios where generated URLs and paths (fully qualified, rooted, relative) would "just work":

1. Re-configuring the proxy to a new path base (e.g. `http://proxy/foo/` -> `http://backend/` to `http://proxy/bar/` -> `http://backend/` ) *without* having to reconfigure and restart the backend server.

2. Having one backend server responding on two separate paths (e.g. `http://proxy/foo/' | `http://proxy/bar/` -> `http://backend/`). Useful if "migrating" to a new route without breaking clients in situ.

**Comment by damianh** [comment]
@Tratcher This is what I use near top of my `Configure(IApplicationBuilder app)`:

            var forwardedHeadersOptions = new ForwardedHeadersOptions
            {
                ForwardedHeaders = ForwardedHeaders.All
            };
            app.UseForwardedHeaders(forwardedHeadersOptions);
            app.Use((context, next) => {
                if (context.Request.Headers.TryGetValue("X-Forwarded-PathBase", out var pathBases))
                {
                    context.Request.PathBase = pathBases.First();
                }
                return next();
            });
